source/common/config/unified_mux/delta_subscription_state.cc

```diff
@@ -0,0 +1,233 @@
+
+#include "envoy/event/dispatcher.h"
+#include "envoy/service/discovery/v3/discovery.pb.h"
+
+#include "common/common/assert.h"
+#include "common/common/hash.h"
+#include "common/config/utility.h"
+#include "common/runtime/runtime_features.h"
+
+namespace Envoy {
+namespace Config {
+namespace UnifiedMux {
+
+DeltaSubscriptionState::DeltaSubscriptionState(std::string type_url,
+                                               UntypedConfigUpdateCallbacks& watch_map,
+                                               std::chrono::milliseconds init_fetch_timeout,
+                                               Event::Dispatcher& dispatcher, const bool wildcard)
+    : SubscriptionState(std::move(type_url), watch_map, init_fetch_timeout, dispatcher),
+      wildcard_(wildcard) {}
+
+DeltaSubscriptionState::~DeltaSubscriptionState() = default;
+
+void DeltaSubscriptionState::updateSubscriptionInterest(
+    const absl::flat_hash_set<std::string>& cur_added,
+    const absl::flat_hash_set<std::string>& cur_removed) {
+  for (const auto& a : cur_added) {
+    resource_state_[a] = ResourceState::waitingForServer();
+    // If interest in a resource is removed-then-added (all before a discovery request
+    // can be sent), we must treat it as a "new" addition: our user may have forgotten its
+    // copy of the resource after instructing us to remove it, and need to be reminded of it.
+    names_removed_.erase(a);
+    names_added_.insert(a);
+  }
+  for (const auto& r : cur_removed) {
+    resource_state_.erase(r);
+    // Ideally, when interest in a resource is added-then-removed in between requests,
+    // we would avoid putting a superfluous "unsubscribe [resource that was never subscribed]"
+    // in the request. However, the removed-then-added case *does* need to go in the request,
+    // and due to how we accomplish that, it's difficult to distinguish remove-add-remove from
+    // add-remove (because "remove-add" has to be treated as equivalent to just "add").
+    names_added_.erase(r);
+    names_removed_.insert(r);
+  }
+}
+
+// Not having sent any requests yet counts as an "update pending" since you're supposed to resend
+// the entirety of your interest at the start of a stream, even if nothing has changed.
+bool DeltaSubscriptionState::subscriptionUpdatePending() const {
+  return !names_added_.empty() || !names_removed_.empty() ||
+         !any_request_sent_yet_in_current_stream_ || dynamicContextChanged();
+}
+
+UpdateAck DeltaSubscriptionState::handleResponse(
+    const envoy::service::discovery::v3::DeltaDiscoveryResponse& response) {
+  // We *always* copy the response's nonce into the next request, even if we're going to make that
+  // request a NACK by setting error_detail.
+  UpdateAck ack(response.nonce(), type_url());
+  TRY_ASSERT_MAIN_THREAD { handleGoodResponse(response); }
+  END_TRY
+  catch (const EnvoyException& e) {
+    handleBadResponse(e, ack);
+  }
+  return ack;
```

Comment:
Should be refactored and implemented at the parent as the SotW variant should do the same thing.

Reply:
done